File: database_connection/task_list/task_list_python/database/db_connection.py
from abc import ABC, abstractmethod


class DbConnection(ABC):
  create_table_query = ""

  def __init__(self) -> None:
    super().__init__()
    self.create_table()

  def create_table(self):
    def operation(_, connection):
      connection.commit()
    self._do_operation(self.create_table_query, operation)

  @abstractmethod
  def get_connection(self):
    pass

  def execute_query(self, query):
    def operation(_, connection):
      connection.commit()
    self._do_operation(query, operation)

  def execute_and_fetch_query(self, query, all=True):
    def operation(cursor, _):
      return cursor.fetchall() if all else cursor.fetchone()
    return self._do_operation(query, operation)

  def _do_operation(self, query, operation):
    try:
      connection = self.get_connection()
      cursor = connection.cursor()
      cursor.execute(query)
      return operation(cursor, connection)
    except Exception as e:
      print("Error on SQL: ", e)
    finally:
      cursor.close()
      connection.close()

File: database_connection/task_list/task_list_java/src/main/java/com/database/DbConnection.java
package com.database;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public abstract class DbConnection implements AutoCloseable {

  protected String driverClassName;
  protected String url;
  protected String user;
  protected String password;
  protected String createTableQuery;
  protected Connection con = null;

  protected abstract void init();

  public DbConnection() {
    init();
    connect();
    createTable();
  }

  public void connect() {
    try {
      Class.forName(driverClassName);
      con = DriverManager.getConnection(url, user, password);
    } catch (SQLException ex) {
      System.out.format("Erro ao se conectar no database: %s \n %s \n %s \n", ex.getSQLState(), ex.getErrorCode(),
          ex.getMessage());
    } catch (Exception ex) {
      System.out.println("Erro ao se conectar no database: " + ex.getMessage());
    }
  }

  @Override
  public void close() {
    try {
      if (con != null) {
        con.close();
      }
    } catch (SQLException ex) {
      System.err.println("Erro ao fechar a conexão: " + ex.getMessage());
    }
  }

  private void createTable() {
    try (var preparedStatement = con.prepareStatement(createTableQuery)) {
      preparedStatement.execute();
    } catch (SQLException ex) {
      System.err.println("Erro ao criar a tabela: " + ex.getMessage());
    }
  }

  public Connection getCon() {
    return con;
  }

  public boolean hasDateTimeType() {
    return !driverClassName.contains("sqlite");
  }
}
